irr/src/CColorConverter.cpp

```diff
@@ -7,6 +7,14 @@
 #include "os.h"
 #include "irrString.h"
 
+// FIXME much of this code does not make sense when endianness is taken into account.
+// It seems likely that the author(s) only thought in terms of big endian,
+// because then byte order and endianness order are the same:
+// The most significant byte is also the first byte.
+// Some of the conversion functions here are thus on little endian inconsistent
+// when it comes to byte order vs. endianness order,
```

Comment:
I can't make sense of these terms in context.
when you have `A8R8G8B8` it is laid out in memory like `[A][R][G][B]`, right?
now if you treat that as a number whether it is `0xAARRGGBB` or `0xBBGGRRAA` depends on endianness.
but what is there to convert?

Reply:
"Endian(ness) order" is most significant byte to least significant byte to me.

"Memory" or "byte" order is just how it lies in memory.

> when you have [ARGB8] it is laid out in memory like [A][R][G][B], right?

In a sane world, ARGB8 *would* mean *either* one of two things:

* `ARGB` in terms of byte order, so `[A][R][G][B]`, no matter whether we're on little endian or big endian.
* `ARGB` in terms of endianness order, so `0xAARRGGBB` (in hindsight this notation would've been clearer).

In the Irrlicht world, it's a mess - both definitions are used; `CColorConverter.cpp` is an especially bad offender - and this comment tries to prepare the reader for that.

Consider for example `convert_B8G8R8toA8R8G8B8`: It treats its *input* as a byte array, so memory order - but it produces its output as an integer using bitops, so endianness order! It converts byte order BGR8 to endianness order ARGB8.

Consider on the other hand `convert_B8G8R8A8toA8R8G8B8`. This just swaps the bytes. So it turns *memory order* BGRA8 into *memory order* ARGB8. But Irrlicht doesn't want *memory order* ARGB8 - because of how `CImage` and most of the code surrounding ARGB8 textures is implemented, it *usually* wants endianness order ARGB8. That is ultimately what caused this bug.

Reply:
I would explain it somewhat like this:
```
// Warning: The naming of Irrlicht color formats is not consistent regarding actual component order in memory
// e.g. ECF_B8G8R8 is handled per-byte and stored as [R][G][B] in memory.
// while ECF_A8R8G8B8 is handled as an u32 0xAARRGGBB so [B][G][R][A] (little endian) in memory
// This also means many conversions are broken on big endian.
```
(illustrative purpose only, details are mostly made up)